Feat: Show titles of GitHub issues on hover fixes: #19710.

brijsiyag opened the pull request:
<!-- Describe your pull request here.-->
Fixes: #19710

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

brijsiyag commented:
@zulipbot add "in progress"

andersk commented on zerver/lib/markdown/__init__.py:
```diff
@@ -1605,6 +1618,8 @@ def url_to_a(
     href = rewrite_local_links_to_relative(db_data, href)
 
     a.set("href", href)
+    if is_previewable(href):
+        a.set("class", "previewable")
```
The results of backend Markdown processing are saved to the database and fixed for all time, so this will prevent us from adding or removing `previewable` in historical messages. This should be done on the frontend, not the backend.

brijsiyag commented:
@andersk Could you please verify the approach in the frontend?
Please ignore tests and commits(messages) for now as I'm working on tests so will handle all these at last.
I have checked the URL on hover only if it is previewable or not and assigned class previewable to it or set attribute title as before. 
Thank you!!

alya commented:
This PR is not running in the dev environment for me, so I'm removing the "product review" label pending errors being fixed. Please post a comment when it's ready for review.

alya commented:
Nice, super excited to see progress on this feature! A few questions/comments after playing with the PR:

1. Let's drop the "Loading" tooltip; I think we're better off not showing a tooltip at all while the preview loads.
2. I noticed that issue <-> pull request redirects are not being handled. E.g. `https://github.com/zulip/zulip/pull/20213` (which is actually an issue) produces an "Invalid or expired URL" tooltip. `https://github.com/zulip/zulip/issue/22551` (which is actually a PR) doesn't produce a custom tooltip at all. Is it possible to fix this? It will be a major problem for linkifiers, since they'll encode both issues and PRs the same way.
3. Is it possible to make this feature work for named links (e.g. `[foo](https://github.com/zulip/zulip/pull/22368)`) as well?

Thanks!

brijsiyag commented:
Okay, I will remove the "Loading" tooltip.
> 2\. tooltip. `https://github.com/zulip/zulip/issue/22551` (which is actually a PR)

This Link is showing 404 on opening, I think you have missed the 's' in issue word.(should be https://github.com/zulip/zulip/issues/22551)
I will check that if GitHub will provide data for redirect urls or not.

and Thanks for letting me know the corner case of named links.
I will try to show tooltips for named links.
Thank you !!

alya commented:
Ah, sorry about the typos when testing -- my bad! OK, I think the only two actual issues to look into are:

1. Let's drop the "Loading" tooltip; I think we're better off not showing a tooltip at all while the preview loads.
2. I noticed that issue <-> pull request redirects are not being handled. E.g. `https://github.com/zulip/zulip/pull/20213` (which is actually an issue) produces an "Invalid or expired URL" tooltip.

timabbott commented on static/js/util.js:
```diff
@@ -246,6 +246,15 @@ export function convert_message_topic(message) {
+        /(?:(?:https?:\/\/)?(?:www.)?)?github.com\/([^/]*)\/([^/]*)\/(issues|pull)\/([^#/?^|]*)/;
+    if (github_re.test(href)) {
+        return true;
+    }
```
Let's replace this regular expression with actually parsing the URL using the URL library. See https://developer.mozilla.org/en-US/docs/Web/API/URL.

timabbott commented on static/js/util.js:
```diff
@@ -284,44 +293,69 @@ export function clean_user_content_links(html) {
-                elt.removeAttribute("title");
-            }
+export function initialize() {
+    $("#main_div").on("mouseenter", "a", (e) => {
```
I think the right selector should be something like `$("body").on("mouseenter", ".rendered_markdown a",` -- this will make this work naturally in stream descriptions, for example. 

(Actually probably we need a tighter class).

timabbott commented on static/js/util.js:
```diff
@@ -284,44 +293,69 @@ export function clean_user_content_links(html) {
+            return;
+        }
+        if (is_url_previewable(href)) {
+            elt.classList.add("previewable");
```
Could you do a `return` here rather than intending all the code below via putting it in a giant `else` clause?

timabbott commented on static/templates/previewable_url_tooltip.hbs:
```diff
@@ -0,0 +1,33 @@
+    .previewable-details {
+        color: gray;
+    }
+</style>
```
Let's put these styles as a section inside `rendered_markdown.css`; we don't do inline styles inside template files as a coding pattern in this project.

timabbott commented on static/templates/previewable_url_tooltip.hbs:
```diff
@@ -0,0 +1,33 @@
+    <img src="/static/images/{{ icon_path }}"/>
+    <div>
+        <div class="previewable-title">{{ title }}</div>
+        <span class="previewable-details">{{ owner }}/{{ repo }}#{{ issue_number }} opened by {{ user }}</span>
```
These strings need to be tagged for translation.

See https://zulip.readthedocs.io/en/latest/translating/internationalization.html to learn how to do this correctly... though I think probably we just want to construct the strings inside JavaScript using `$t`, since that'll provide a more flexible interface if we want to support things that are not GitHub/GitLab in the future. So I'd just make the parameters be `hover_preview_icon_path`, `hover_preview_title`, `hover_preview_details`.

timabbott commented on zerver/lib/github.py:
```diff
@@ -61,3 +66,44 @@ def get_latest_github_release_download_link_for_platform(platform: str) -> str:
+def get_issue_or_pr_cache_key(
+    owner: str, repo: str, type: str, issue_number: str
+) -> str:  # nocoverage
+    return f"github_issue_or_pr_url:{owner}/{repo}/{type}/{issue_number}"
```
Please add some tests that cover this code path. You can do it nicely using the `responses` module to configure what HTTP responses GitHub is expected to return from a given API endpoint; this is important for the tests to work offline and without a real API key.

(Often we setup a fixture)

timabbott commented on zerver/lib/github.py:
```diff
@@ -61,3 +66,44 @@ def get_latest_github_release_download_link_for_platform(platform: str) -> str:
+    try:
+        response = GithubSession().get(api_url)
+        response.raise_for_status()
+        json_data: Dict[str, Any] = response.json()
```
I think `json_data` is a misleading name for this variable -- it's just a dictionary at this point, not a JSON string, having been decoded from the original JSON.

timabbott commented on zerver/views/previewable.py:
```diff
@@ -0,0 +1,29 @@
+    request: HttpRequest, user_profile: UserProfile, url: str = REQ()
+) -> HttpResponse:
+    github_re = r"(?:(?:https?:\/\/)?(?:www.)?)?github.com\/([^\/]*)\/([^\/]*)\/(issues|pull)\/([^\/|^?|#]*)"
+    github_match = re.match(github_re, url)
```
Same comment on parsing here; we should let `urrlib` parse the library and then just compare the hostname to github.com, for example. This is much safer/more secure and easier to read too :)

brijsiyag replied:
@timabbott, but how we will be knowing if it is a user/org URL or an Issue/PR URL?
Same doubt in the above comment for `util.js`.

timabbott replied:
We'll still need to look at the URL's `path` component, probably with a regex -- i.e. the part that isn't the domain. What I'm trying to do is have the `https` and `github.com` parts use a standard URL library and then just check "is it an https URL {"github.com", "www.github.com"} with clean JS given that's what we mean. The last bit of just parsing the path will still need to involve a regex, of course, but it'll just be on the URL's path -- the `?` parameters, `#` fragments, and other complexity will be handled for us.

timabbott commented:
Thanks for working on this @brijsiyag! I posted a batch of mostly tactical things that you should be able to fix.

brijsiyag commented:
@zulipbot add in progress

brijsiyag commented:
@zulipbot remove "has conflicts"

brijsiyag commented:
Hey @timabbott , I have addressed all the reviews. PTAL.
Screen capture:-

https://user-images.githubusercontent.com/71015892/227720078-b3239497-459a-499f-a7a3-b2ae8d28ebca.mov

alya commented:
I did another round of manual testing. Mostly it looks great, though we still have the issue that if you have, e.g., a `/pull` URL instead of an `/issues` URL, you get an "Invalid or expired URL." tooltip. My main concern with this is that it will come up all the time when using linkifiers.

brijsiyag commented:
> I did another round of manual testing. Mostly it looks great, though we still have the issue that if you have, e.g., a `/pull` URL instead of an `/issues` URL, you get an "Invalid or expired URL." tooltip. My main concern with this is that it will come up all the time when using linkifiers.

@alya , Sorry I missed the earlier comment while working on dropping `Loading...` text.
I have updated the code and attaching screen capture below testing all combinations of states with issues and pull requests.
Thank you !!

https://user-images.githubusercontent.com/71015892/227953841-f49a3838-d07c-4476-a164-26d087794cb6.mov

alexmv commented on zerver/tests/test_previewable.py:
```diff
@@ -0,0 +1,140 @@
+            "locked": False,
+            "state_reason": None,
+        }
+        responses.add(
+            responses.GET, "https://api.github.com/repos/zulip/zulip/issues/1", json=response_data
+        )
```
Shouldn't the cache mean that we _don't_ need this response?

brijsiyag replied:
Ooh, I missed that, will address this in next commit.

alexmv replied:
If the cache is working, then it *shouldn't* see the new data in the response, because it should return the cached data from the first response, and not make a second request.

brijsiyag replied:
Okay, I will update this to like we will first add a mock response than request for data than remove the mock response and than again fetch the data and assert.

alexmv commented on zerver/tests/test_previewable.py:
```diff
@@ -0,0 +1,140 @@
+
+    @responses.activate()
+    def test_get_github_pull_data(self) -> None:
+        response_data = {
+            "title": "Fix 'python-django-guardian' download",
+            "user": {"login": "littldr"},
+            "state": "open",
+            "pull_request": {
+                "merged_at": None,
+            },
+        }
```
We may want to use a more real fixture for this.

alexmv commented on zerver/lib/github.py:
```diff
@@ -61,3 +66,42 @@ def get_latest_github_release_download_link_for_platform(platform: str) -> str:
+        data.update(additional_data)
+        return data
+    except requests.RequestException as err:
+        logger.exception("Unable to fetch data from GitHub %s", api_url, stack_info=True)
```
I think we should be a little more careful with the log level here, based on the status code:
 - 5xx error codes: there's nothing the server administrator can do, so filling out the Zulip server error logs isn't going to help -- we should use INFO at most.
 - 404 seem like a user error, so I don't think we need to log it at any level.
 - 403s are possibly a server error (if they're expecting it to be able to access a private repo but have the token misconfigured) but in larger deploys it's not something worth logging over, since users can enter arbitrary information.
 - 401s are probably a WARNING or ERROR, since that's something wrong with the token itself
 - 429s are a WARNING, since fundamentally it can be caused by users taking bad actions -- but something that affects the whole server so the admin should be more aware.

In none of these should we use `exception` since I don't think the stacktrace is useful.

brijsiyag replied:
I have checked on rate limit status code is `403 rate limit exceeded` for Github api, this is the one we should sure show?

brijsiyag replied:
Yes, I have showed warning on 401, 403 and 429.

alexmv commented on zerver/views/previewable.py:
```diff
@@ -0,0 +1,34 @@
+def get_previewable_data(
+    request: HttpRequest, user_profile: UserProfile, url: str = REQ()
+) -> HttpResponse:
+    parsed_url = urlparse(url)
```
`urlparse` can fail (e.g. with `https://boom[here`), which we should catch gracefully rather than 500.

brijsiyag replied:
I went through codebase, we are not returning 422(Unprocessable Content) status code anywhere. Should we consider this for these type of errors?

alexmv commented on zerver/views/previewable.py:
```diff
@@ -0,0 +1,34 @@
+) -> HttpResponse:
+    parsed_url = urlparse(url)
+    if (
+        parsed_url.hostname
+        and re.match(r"^(?:www\.)?github\.com$", parsed_url.hostname) is not None
```
This accepts to preview `ftp://www.github.com/` and `http://www.github.com:9000/`, which it shouldn't.

alexmv commented on zerver/views/previewable.py:
```diff
@@ -0,0 +1,34 @@
+        parsed_url.hostname
+        and re.match(r"^(?:www\.)?github\.com$", parsed_url.hostname) is not None
+    ):
+        path_split = parsed_url.path.split("/")
+        if path_split[3] in ["issues", "pull"]:
```
If this gets handed `https://github.com` then `parsed_url` is `""`, so `path_split` is `[""]`, which throws a 500 when accessing `path_split[3]`.

Ditto, later, if one were to pass `https://github.com/foo/bar/issues`.

alexmv commented on zerver/lib/github.py:
```diff
@@ -61,3 +66,42 @@ def get_latest_github_release_download_link_for_platform(platform: str) -> str:
+
+
+def get_issue_or_pr_cache_key(owner: str, repo: str, issue_number: str) -> str:
+    return f"github_issue_or_pr_data:{owner}/{repo}/{issue_number}"
```
This cache key should include a hash of the github token, since the contents of the response from Github also vary by that.  This means that if you change the key, you immediately start seeing responses which are limited to what that new key can access.

We need to use a hash of the github token, since the cache keys shouldn't themselves have the secret embedded in them.

alexmv commented on zerver/views/previewable.py:
```diff
@@ -0,0 +1,34 @@
+            if data["result"] == "success":
+                return json_success(request, data)
+            error_msg = data["msg"]
+            raise JsonableError(f"{_(error_msg)}")
```
This isn't translatable, since it can be an arbitrary string.  The f-string isn't helping improve the readability, either. :)

I'm guessing that you want to cache the failure, which is why `get_issue_or_pr_data` isn't raising the JsonableError directly.  Which is reasonable for 404s -- those are unlikely to want to try to fetch fresh each time. If you're going to do this, I think you need to return some opaque constant from get_issue_or_pr_data, and check those values here and raise JsonableError with the user-facing string explicitly.

alexmv commented on zerver/tests/test_openapi.py:
```diff
@@ -275,6 +275,8 @@ class OpenAPIArgumentsTest(ZulipTestCase):
         # Zulip outgoing webhook payload
         "/zulip-outgoing-webhook",
+        # Api for previewable url
+        "/previewable",
```
We're going to want to document this in the OpenAPI spec eventually, since we'd like any client to be able to use it.

brijsiyag replied:
Added it here because checks were failing, I will add an issue or put a PR after this PR closes.

alexmv reviewed (commented):
I'm not qualified to review the frontend half, but I've left some comments on the first commit.

On the whole, we need to be a bit more defensive about what the URL we're getting looks like, since even if we're constraining it with regex on the client side, it _is_ a public API, and people will use it to do things we don't necessarily expect. :)

brijsiyag commented on zerver/lib/github.py:
```diff
@@ -61,3 +66,42 @@ def get_latest_github_release_download_link_for_platform(platform: str) -> str:
+    return f"github_issue_or_pr_data:{owner}/{repo}/{issue_number}"
+
+
+@cache_with_key(get_issue_or_pr_cache_key, timeout=3600)
```
Okay I will do it for 10 minutes, we can consider increasing the value as per traffic later.

timabbott replied:
Let's also add a comment documenting the trade-off with the 5000 requests/hour rate-limit vs. having current data.

brijsiyag commented:
@zulipbot add "in progress"

brijsiyag commented:
@alexmv , Thanks for the feedback. I have updated the code, ready for next round review.

alexmv commented on zerver/tests/test_previewable.py:
```diff
@@ -0,0 +1,212 @@
+
+    @responses.activate()
+    def test_github_without_auth(self) -> None:
+        def request_callback(request: requests.PreparedRequest) -> Tuple[int, Dict[str, str], str]:
+            if (
+                request.headers.get("Authorization", None)
+                == "token 3608bca1e44ea6c4d268eb6db02260269892c0"
+            ):
+                return (200, {}, json.dumps({}))  # nocoverage
+            return (403, {}, json.dumps({}))
```
You have this code twice, slightly differently -- and in neither case do you actually _need_ the `request_callback`, since in each case you only exercise one side of the `if` statement.

You should pull this helper outside of your two test methods, so it gets exercised.

alexmv commented on zerver/lib/github.py:
```diff
@@ -61,3 +67,38 @@ def get_latest_github_release_download_link_for_platform(platform: str) -> str:
+            else None
+        )
+        if status_code in [401, 403, 429]:
+            logger.warning("Unable to fetch data from GitHub %s", str(err))
```
`Unable to fetch data from GitHub 403 Client Error: Forbidden for url: https://...` isn't terribly readable -- we want some separator before the `err`, if only a `:`:
```suggestion
            logger.warning("Unable to fetch data from GitHub: %s", str(err))
```

I also wonder if we should avoid caching this set of results, by raising an exception.  Thoughts, @timabbott?

brijsiyag replied:
Currently, the code caches this information, but we will update it once we make a decision.

alexmv commented on zerver/tests/test_previewable.py:
```diff
@@ -0,0 +1,212 @@
+
+    @responses.activate()
+    def test_invalid_previewable_url(self) -> None:
+        response_data = (
+            {
+                "message": "Not Found",
+                "documentation_url": "https://docs.github.com/rest/reference/issues#get-an-issue",
+            },
+        )
```
I don't believe this needs to be a tuple.

brijsiyag replied:
`lint --fix` did this change based on `,` I think, updated it.

alexmv commented on zerver/tests/test_previewable.py:
```diff
@@ -0,0 +1,212 @@
+        response = self.client_post(
+            "/json/previewable", {"url": "https://github.com/zulip/zulip/issues/2"}
+        ).json()
+        self.assertEqual(response["msg"], "Invalid or expired URL.")
```
I think we should use `self.assert_json_error` here, so we also check the status code.

alexmv commented on zerver/tests/test_previewable.py:
```diff
@@ -0,0 +1,212 @@
+        self.client_post(
+            "/json/previewable", {"url": "https://github.com/zulip/zulip/issues/19710"}
+        )
+        responses.remove(responses.GET, "https://api.github.com/repos/zulip/zulip/issues/19710")
```
Rather than remove it, you can use `responses.assert_call_count(...)`

alexmv commented on zerver/tests/test_previewable.py:
```diff
@@ -0,0 +1,212 @@
+        result = self.client_post(
+            "/json/previewable", {"url": "https://github.com/zulip/zulip/issues/19710"}
+        ).json()
+        self.assertEqual(
+            result["title"],
+            "Show titles of GitHub issues on hover",
+        )
+        self.assertEqual(
+            result["user"],
+            {"login": "brijsiyag"},
+        )
+        self.assertEqual(
+            result["state"],
+            "open",
+        )
+        self.assertEqual(
+            result["locked"],
+            False,
+        )
+        self.assertEqual(
+            result["state_reason"],
+            None,
+        )
```
These don't add much to the test, since we test that above.  What seems more interesting is that the first and second responses were both successful, and have the same body, and that the second one didn't cause an HTTP request:
```python3
        first_response = self.client_post("/json/previewable", {"url": url})
        self.assert_json_success(first_response)
        responses.assert_call_count(api_url, 1)

        second_response = self.client_post("/json/previewable", {"url": url})
        self.assert_json_success(second_response)
        responses.assert_call_count(api_url, 1)

        self.assertEqual(first_response.json(), second_response.json())

alexmv commented on zerver/lib/github.py:
```diff
@@ -12,7 +14,11 @@
 class GithubSession(OutgoingSession):
     def __init__(self, **kwargs: Any) -> None:
-        super().__init__(role="github", timeout=5, **kwargs)
+        github_api_auth_token = settings.GITHUB_API_AUTH_TOKEN
```
You should add a commented section to `zproject/prod_settings_template.py` (search for "Previews") which doesn't have any settings, but does describe that this is a secret that can be set, and how to generate the token (e.g. can one use [fine-grained access tokens](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token#fine-grained-personal-access-tokens), and is so what rights are needed?).

As written, it's not *required* to get previews, but setting it will allow users to make better use of them -- e.g. describing that it can give access to private repos seems like a good thing to note and will make folks potentially more interested in the feature.  Mentioning what the quotas are with and without a token would also be good.

alexmv reviewed (commented):
This is shaping up nicely, @brijsiyag!

timabbott commented on web/src/tippyjs.js:
```diff
@@ -382,6 +385,74 @@ export function initialize() {
+                                hover_preview_details: $t(
+                                    {
+                                        defaultMessage:
+                                            "{ owner }/{ repo }#{ issue_number } opened by { user }",
```
I think it'd be nicer to construct a `{issue}` variable for the first part, and have this be spelled for translations `"{issue} opened by {user}"`. (We usually don't put extra space around variable names).

timabbott commented on web/src/util.ts:
```diff
@@ -302,45 +319,74 @@ export function clean_user_content_links(html: string): string {
-                title = url.toString();
-                legacy_title = href;
+                // For non-image user uploads, the following block ensures that the title
+                // attribute always displays the filename as a security measure.
```
Can you comment on the security implications of moving this out of `clean_user_content_links`? Is the idea that we're no longer using `title` attributes or something?

brijsiyag replied:
I can't think of any security concerns here as we are just moving the setting title part out of the `clean_user_content_links`, as we have to decide whether to show a title or not based on whether URL is previewable or not.

timabbott commented on zerver/views/previewable.py:
```diff
@@ -0,0 +1,45 @@
+) -> HttpResponse:
+    try:
+        parsed_url = urlparse(url)
+    except:  # noqa: E722
```
More generally, we almost never as a project want to be suppressing linter errors without actually addressing the feedback from them; we try pretty hard to tune our linters down to fairly precise rules.

If you run into something like this in the future, it's best to post about it in development help and we can help you figure out the right way to resolve it.

timabbott commented:
Dropping this from the list of release goals, since it's not on track to merge in time for 7.0.

brijsiyag commented:
@timabbott , if you got some time review it?
